distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala

```diff
@@ -962,6 +967,64 @@ object Lifecycle extends LifecycleCatsInstances {
     }
   }
 
+  @inline
+  private final def foldMImpl[F[_], A, B](
+    self: Lifecycle[F, A]
+  )(failure: Throwable => Lifecycle[F, B],
+    success: A => Lifecycle[F, B],
+  )(implicit F: QuasiIO[F]
+  ): Lifecycle[F, B] = {
+    import QuasiIO.syntax._
+    new Lifecycle[F, B] {
+      override type InnerResource = AtomicReference[List[() => F[Unit]]]
+
+      private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
```

Comment:
```suggestion
      private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: => Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
```

Must suspend here to ensure that any effects in construction of lifecycle are wrapped into `bracket`'s `try` (in Identity instance). (Lifecycle/QuasiIO must work correctly with impure code)

Reply:
Hmm, this may be tricky because of path dependency

Reply:
@VladPodilnyk 
This seems to work as an encoding that supports suspension and path dependency:
```scala

      private[this] sealed trait LifecyclePair[+a] {
        val lifecycle: Lifecycle[F, a]
        val inner: lifecycle.InnerResource
      }
      private[this] def LifecyclePair[a](lifecycle0: Lifecycle[F, a])(inner0: lifecycle0.InnerResource): LifecyclePair[a] = new LifecyclePair[a] {
        override val lifecycle: lifecycle0.type = lifecycle0
        override val inner: lifecycle.InnerResource = inner0
      }

      private[this] def bracketAppendFinalizer[a, b, t](
        finalizers: InnerResource
      )(lifecycleCtor: () => Lifecycle[F, a]
      )(use: LifecyclePair[a] => F[b]
      ): F[b] = {
        F.bracket(
          acquire = {
            val lifecycle = lifecycleCtor()
            lifecycle.acquire.flatMap {
              a =>
                F.maybeSuspend {
                  // can't use `.updateAndGet` because of Scala.js
                  var oldValue = finalizers.get()
                  while (!finalizers.compareAndSet(oldValue, (() => lifecycle.release(a)) :: oldValue)) {
                    oldValue = finalizers.get()
                  }
                  LifecyclePair(lifecycle)(a)
                }
            }
          }
        )(release = _ => F.unit)(
          use = use
        )
      }
```

Reply:
Do you propose to extract it into a pattern ☝️ ?
(it's not clear for me what problem does it solves)
Also, we probably should come up with another name bc this one is not very informative, IMO.

Reply:
@VladPodilnyk 
> (it's not clear for me what problem does it solves)

You need the `lifecycle` parameter to be by-name because it may produce effects immediately when called, instead of when executed (it's natural to perform effects inplace if you use impure Identity instance)

But if you write it as by-name, it's now impossible to write the `lifecycle.InnerResource` type:
```scala
private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: => Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
```

This type will produce an error because `lifecycle` is now by-name - every time you refer to it in code, the by-name parameter will run again and produce a new value (operationally `=> Lifecycle[F, a]` is exactly the same as `() => Lifecycle[F, a]`). 
Because it produces a new lifecycle each time, you cannot refer to `lifecycle.InnerResource` — a type member of `lifecycle` value — because there is no "`lifecycle` value" here, only a function that makes new Lifecycles. So we package the result "value" into a trait to make it referable (the reason it's not a class is because classes do not support path-dependency in Scala 2, only traits)

In short, try to implement the above signature and you'll be able to experience the limitations that force it to be this way.

Reply:
Right now you need to get rid of `val` extraction and write this thing as if everything is imperative (because it IS when using `Identity`) and then it's good to merge 🚀

Reply:
@neko-kai, thanks for this excellent explanation!